python/nano/src/bigdl/nano/deps/ray/ray_distributed.py

```diff
@@ -72,49 +79,103 @@ def get_env_vars(self, key: str):
-    If using this plugin, you should run your code like a normal Python
-    script: ``python train.py``, and only on the head node if running in a
-    distributed Ray cluster. There is no need to run this script on every
-    single node.
-    Args:
-        num_workers (int): Number of training workers to use.
-        num_cpus_per_worker (int): Number of CPUs per worker.
-        use_gpu (bool): Whether to use GPU for allocation. For GPU to be
-            used, you must also set the ``gpus`` arg in your Pytorch Lightning
-            Trainer to a value > 0.
-        init_hook (Callable): A function to run on each worker
-            upon instantiation.
-        **ddp_kwargs: Additional arguments to pass into
-            ``DistributedDataParallel`` initialization
-    Example:
-        .. code_block:: python
-            import pytorch_lightning as ptl
-            from ray_lightning import RayAccelerator
-            ptl_model = MNISTClassifier(...)
-            plugin = RayPlugin(num_workers=4, cpus_per_worker=1,
-                use_gpu=True)
-            # If using GPUs, set the ``gpus`` arg to a value > 0.
-            # The actual number of GPUs is determined by ``num_workers``.
-            trainer = pl.Trainer(..., gpus=1, plugins=[plugin])
-            trainer.fit(ptl_model)
-    """
+class _RayLauncher(_SpawnLauncher):
+
+    def __init__(self, strategy: Strategy) -> None:
+        self._strategy = strategy
+
+    @property
+    def is_interactive_compatible(self) -> bool:
+        return False
+
+    def launch(self, function: Callable, *args: Any,
+               trainer: Optional["pl.Trainer"] = None, **kwargs: Any) -> Any:
+        # pytorch_lightning 1.6 uses this method to create child processes
+        strategy = self._strategy
+
+        # fix bug
+        # args[1] is dataloader, args[3] and args[4] is datamodule
+        if self._strategy.use_ipex and TORCH_VERSION_LESS_1_10:
+            if isinstance(args[1], LightningDataModule):
+                args[1].trainer = None
+            elif isinstance(args[3], LightningDataModule):
+                args[3].trainer = None
+            elif isinstance(args[4], LightningDataModule):
+                args[4].trainer = None
+
+        strategy._setup_env_vars()
+        strategy.global_to_local = strategy.get_local_ranks()
+
+        torch_backend = os.getenv("PL_TORCH_DISTRIBUTED_BACKEND")
+        if torch_backend is None:
+            torch_backend = "nccl" if strategy.use_gpu else "gloo"
+        strategy._process_group_backend = torch_backend
+
+        futures = [
+            strategy.workers[i].execute.remote(
+                self._wrapping_function,
+                (i, trainer, function, args, kwargs)
+            )
+            for i in range(strategy.num_workers)
+        ]
+
+        not_ready = futures
+        while not_ready:
+            ready, not_ready = ray.wait(not_ready, timeout=0)
+            ray.get(ready)
+        ray.get(ready)
```

Comment:
I think these lines are equivalent to `ray.get(futures)`, right?

Reply:
these lines are copied from the original `execution_loop` method, but it seems they are redundant, I'll delete them